Feat/mech integration

axatbhardwaj opened the pull request:
Mech interact works 

currently only uses openai-gpt-3.5-turbo

dvilelaf commented on packages/dvilela/skills/memeooorr_abci/behaviour_classes/chain.py:
```diff
@@ -30,6 +30,9 @@
 from packages.dvilela.skills.memeooorr_abci.behaviour_classes.base import (
     MemeooorrBaseBehaviour,
 )
+from packages.dvilela.skills.memeooorr_abci.behaviour_classes.twitter import (
+    EngageTwitterBehaviour,
+)
```
Why do we need this?

axatbhardwaj replied:
we don't need it anymore removing it

dvilelaf commented on packages/dvilela/skills/memeooorr_abci/behaviour_classes/chain.py:
```diff
@@ -728,6 +735,13 @@ def async_act(self) -> Generator:
             ):
                 event = Event.ACTION.value
 
+            # check for
```
Unfinished comment

axatbhardwaj replied:
done

dvilelaf commented on packages/dvilela/skills/memeooorr_abci/behaviour_classes/mech.py:
```diff
@@ -0,0 +1,134 @@
+                )
+                payload = PostMechRequestPayload(
+                    sender=sender,
+                    mech_for_twitter=False,
```
naming is a bit too specific. Post mech round is for postprocessing mech requests, we shouldnt tie it to twitter

axatbhardwaj replied:
The variable/property is used by Engage twitter round so , kept it specific 

we decide if we need to go into a specific code block via this vraible which is decided by mech interact and other factors

dvilelaf commented on packages/dvilela/skills/memeooorr_abci/behaviour_classes/mech.py:
```diff
@@ -0,0 +1,134 @@
+                    mech_for_twitter=False,
+                )
+            else:
+                self.context.logger.error(
+                    "Mech response Found, the LLM will use the mech response"
```
This behaviour should know anything about any LLM

axatbhardwaj replied:
changing it

dvilelaf commented on packages/dvilela/skills/memeooorr_abci/behaviour_classes/mech.py:
```diff
@@ -0,0 +1,134 @@
+        self.set_done()
+
+
+class FailedMechRequestBehaviour(
```
What do we need this round for?

axatbhardwaj replied:
we need to know if the mech request/response has failed

after discussing it with manos the most efficient way to do was to have degenerate/normal rounds in our ABCI 

we could have had 1 common round but there were multiple cases(reasons) for failures so for different mech tools they may differ , that is why we have these different round 
currently we do not do anything significant just change a boolean property in synchronized data named : mech_for_twitter

axatbhardwaj replied:
![Mech_interact new flow](https://github.com/user-attachments/assets/de5a755c-13e7-4d94-85be-b486f7911eac)

just to provide you more context

dvilelaf commented on packages/dvilela/skills/memeooorr_abci/behaviour_classes/mech.py:
```diff
@@ -0,0 +1,134 @@
+        self.set_done()
+
+
+class FailedMechResponseBehaviour(
```
Same. Why do we need it? It seems we do nothing here

axatbhardwaj replied:
To know exactly where mech interact ABCI failed , pls refer to [composition.py](https://github.com/dvilelaf/meme-ooorr/blob/6fafe9997209e16c396ac2cbd6168a641b66eab1/packages/dvilela/skills/memeooorr_chained_abci/composition.py#L55) for this

dvilelaf commented on packages/dvilela/skills/memeooorr_abci/prompts.py:
```diff
@@ -25,17 +25,28 @@
 
 
-TWITTER_DECISION_PROMPT = """
+TWITTER_DECISION_PROMPT_WITH_TOOLS = """
```
We now have 2 prompts, but 80% of those prompts is the same. Try to extract the common part and conditionally add either the tools or the mech response part. Avoid duplications. Also check capitalization and typos.

axatbhardwaj replied:
Done

dvilelaf commented on scripts/test_prompts_tools.py:
```diff
@@ -0,0 +1,392 @@
+# print(json.loads(response.text))
+
+
+TWITTER_DECISION_PROMPT_WITH_TOOLS = """
```
Are those prompt the same in prompts.py? If so, remove here and import

dvilelaf commented on packages/dvilela/skills/memeooorr_abci/rounds.py:
```diff
@@ -149,11 +159,33 @@ def participant_to_staking(self) -> DeserializedCollection:
         return int(self.db.get("tx_loop_count", 0))  # type: ignore
 
+    @property
+    def mech_for_twitter(self) -> bool:
```
What is this?

axatbhardwaj replied:
It is the boolean that tells, EngagetwitterRound if the llm requested mech in the previous round and this round we need to provide the mech responses

dvilelaf commented on packages/dvilela/skills/memeooorr_abci/behaviour_classes/twitter.py:
```diff
@@ -43,10 +46,15 @@
 from packages.valory.skills.abstract_round_abci.base import AbstractRound
 
 
+TEMP_TOOLS_LIST = """
+openai-gpt-3.5-turbo: This tool generates a tweet based on a given prompt using the OpenAI GPT-3.5-turbo model.
+"""
+
```
Tools shouldnt be a string. It should be a dict. Also, it should be a param

dvilelaf commented on packages/dvilela/skills/memeooorr_abci/behaviour_classes/twitter.py:
```diff
@@ -320,90 +328,161 @@ def async_act(self) -> Generator:
+                mech_requests = json.dumps(new_mech_requests, sort_keys=True)
+                self.context.logger.info(f"Mech Requests JSON: {mech_requests}")
+
+            # add new prop to payload for tx_submitter and check for it in posttxround
+            if event == Event.MECH.value:
+                payload = EngageTwitterPayload(
+                    sender=self.context.agent_address,
+                    event=event,
+                    mech_request=json.dumps(new_mech_requests, sort_keys=True),
+                    tx_submitter=self.matching_round.auto_round_id(),
+                )
+            else:
+                payload = EngageTwitterPayload(
+                    sender=self.context.agent_address,
+                    event=event,
+                    mech_request=None,
+                    tx_submitter=self.matching_round.auto_round_id(),
+                )
```
Check capitalization in comments. Why duplicate the full instance instead of just having a conditional mech_request object?

dvilelaf commented on packages/dvilela/skills/memeooorr_abci/behaviour_classes/twitter.py:
```diff
@@ -320,90 +328,161 @@ def async_act(self) -> Generator:
         self.set_done()
 
-    def get_event(self) -> Generator[None, None, str]:
+    def get_event(  # pylint: disable=too-many-statements
```
This function is enormous now. Always try to split logic in small functions that are easy to read and mantain. It's hard to me to follow the logic. Also reduce nesting as much as possibleor you will start having a lot of corner cases you hadnt imagined.

dvilelaf commented on packages/dvilela/skills/memeooorr_abci/behaviour_classes/twitter.py:
```diff
@@ -417,145 +496,262 @@ def get_event(self) -> Generator[None, None, str]:
-            other_tweets=other_tweets,
-            time=self.get_sync_time_str(),
-        )
+            tweets = yield from self.get_previous_tweets()  # type: ignore
+            tweets = tweets[-5:] if tweets else None  # type: ignore
+
+            if tweets:
+                random.shuffle(tweets)
+                previous_tweets = "\n\n".join(  # type: ignore
+                    [
+                        f"tweet_id: {tweet['tweet_id']}\ntweet_text: {tweet['text']}\ntime: {datetime.fromtimestamp(tweet['timestamp']).strftime('%Y-%m-%d %H:%M:%S')}"
+                        for tweet in tweets
```
In general, this file needs a lot of cleanup. It just looks too dirty. We need encapsulation, good code comments, less nesting. I see a lot of type ignores. Be careful with always ignoring lint errors, we have them for a reason. Always ensure that you only ignore if it makes sense.

axatbhardwaj commented on packages/dvilela/skills/memeooorr_chained_abci/composition.py:
```diff
@@ -45,8 +49,16 @@
+    MemeooorrAbci.FinishedForMechRequestRound: MechRequestStates.MechRequestRound,
+    MechFinalStates.FinishedMechRequestRound: TransactionSettlementAbci.RandomnessTransactionSubmissionRound,
+    MechFinalStates.FinishedMechResponseRound: MemeooorrAbci.PostMechRequestRound,  # Happy path
+    # need to do somehthing about failed states
```
Ignore the comment here , removed it

axatbhardwaj commented on packages/dvilela/skills/memeooorr_abci/rounds.py:
```diff
@@ -321,15 +371,187 @@ def end_block(self) -> Optional[Tuple[BaseSynchronizedData, Event]]:
+                        # Validate mech requests format
+                        if not isinstance(mech_requests, list):
+                            self.context.logger.error("Invalid mech_requests format")
+                            return self.synchronized_data, Event.ERROR
+
+                        # Update synchronized data with new mech requests
+                        synchronized_data = self.synchronized_data.update(
+                            synchronized_data_class=SynchronizedData,
+                            **{
+                                get_name(SynchronizedData.mech_requests): json.dumps(
+                                    mech_requests, cls=DataclassEncoder
+                                ),
+                            },
+                        )
+                        return synchronized_data, Event.MECH
+                    except json.JSONDecodeError as e:
+                        self.context.logger.error(f"Failed to parse mech_request: {e}")
+                        return self.synchronized_data, Event.ERROR
+
+            # If no mech requests, proceed with normal flow
+            try:
+                # Handle any other payload data updates here
+                # changing the mech_for_twitter to False
+                synchronized_data = self.synchronized_data.update(
+                    synchronized_data_class=SynchronizedData,
+                    **{
+                        get_name(SynchronizedData.mech_for_twitter): False,
+                    },
+                )
+                return synchronized_data, Event.DONE
+            except Exception as e:  # pylint: disable=broad-except
+                self.context.logger.error(f"Error processing payload: {e}")
+                return self.synchronized_data, Event.ERROR
+
+        if not self.is_majority_possible(
+            self.collection, self.synchronized_data.nb_participants
+        ):
+            return self.synchronized_data, Event.NO_MAJORITY
+
+        return None
```
here i need to evaluate mech requests json and put them in the synchronised data , a similar patter was observed in [IEKIT](https://github.com/valory-xyz/IEKit/blob/e0d9daefebe953a9aa9d26eb9b47680daa809dd0/packages/valory/skills/twitter_scoring_abci/rounds.py#L494-L495) as well

dvilelaf replied:
Dont do that, trust me. IEKit was one of out initial agent architecture and you dont want to follow that pattern.

axatbhardwaj commented:
- reworked Engage Twitter behavior
  - 1. get event function
  - 2. Twitter interact function 
  - both were split into smaller units of logical functions  
- Prompt redundancy was removed 
- tools are added as params 
- reduced linter-check skips by 80%
- cleaned up a code a lot from behaviors to composition

Adamantios commented on packages/dvilela/skills/memeooorr_abci/payloads.py:
```diff
@@ -115,6 +117,27 @@ class CallCheckpointPayload(BaseTxPayload):
     tx_hash: Optional[str]
 
 
+@dataclass(frozen=True)
+class PostMechRequestPayload(BaseTxPayload):
+    """Represent a transaction payload for the PostMechRequestRound."""
+
+    mech_for_twitter: bool
+
+
+@dataclass(frozen=True)
+class FailedMechResponsePayload(BaseTxPayload):
+    """Represent a transaction payload for the PostMechResponseRound."""
+
+    mech_for_twitter: bool
+
+
+@dataclass(frozen=True)
+class FailedMechRequestPayload(BaseTxPayload):
+    """Represent a transaction payload for the PostMechResponseRound."""
+
+    mech_for_twitter: bool
+
+
```
No need to have 3 payloads with the same content; could only use one of them for all the rounds.

Adamantios commented on packages/dvilela/skills/memeooorr_abci/rounds.py:
```diff
@@ -149,11 +159,33 @@ def participant_to_staking(self) -> DeserializedCollection:
         """Get the participants to the staking round."""
         return self._get_deserialized("participant_to_staking")
 
+    @property
+    def mech_requests(self) -> List[MechMetadata]:
+        """Get the mech requests."""
+        serialized = self.db.get("mech_requests", "[]")
+        if serialized is None:
+            serialized = "[]"
+        requests = json.loads(serialized)
+        return [MechMetadata(**metadata_item) for metadata_item in requests]
+
+    @property
+    def mech_responses(self) -> List[MechInteractionResponse]:
+        """Get the mech responses."""
+        responses = self.db.get("mech_responses", "[]")
+        if isinstance(responses, str):
+            responses = json.loads(responses)
+        return [MechInteractionResponse(**response_item) for response_item in responses]
+
```
Why not override the synced data of the mech interact? I think it contains these properties.

axatbhardwaj replied:
because David has already implemented a base behavior for the agent and I didn't want to do multiple inheritance or change the base behavior for the agent as I needed only two properties , I will keep this in mind and discuss it with him once he's back

@Adamantios

Adamantios replied:
Beahviour is not related to the synchronized data.

Adamantios commented on packages/dvilela/skills/memeooorr_abci/rounds.py:
```diff
@@ -321,15 +371,187 @@ def end_block(self) -> Optional[Tuple[BaseSynchronizedData, Event]]:
+        return None
+
+
+class FailedMechResponseRound(CollectSameUntilThresholdRound):
+    """FailedMechResponseRound handles the case where the mech response is not received.it is always going to end in EngageTwitterRound"""
+
+    payload_class = FailedMechResponsePayload
+    synchronized_data_class = SynchronizedData
+    extended_requirements = ()
+
+    def end_block(self) -> Optional[Tuple[BaseSynchronizedData, Event]]:
+        """Process the end of the block."""
+        if self.threshold_reached:
+            # This needs to be mentioned for static checkers
+            # Event.DONE, Event.NO_MAJORITY, Event.ROUND_TIMEOUT, Event.ERROR
+            payload = FailedMechResponsePayload(
+                *(("dummy_sender",) + self.most_voted_payload_values)
+            )
+
+            self.context.logger.info(
+                f"FailedMechResponseRound payload recived: {payload}"
+            )
+
+            synchronized_data = self.synchronized_data.update(
+                synchronized_data_class=SynchronizedData,
+                **{
+                    get_name(
+                        SynchronizedData.mech_for_twitter
+                    ): payload.mech_for_twitter,
+                },
+            )
+
+            return synchronized_data, Event.DONE
+
+        if not self.is_majority_possible(
+            self.collection, self.synchronized_data.nb_participants
+        ):
+            return self.synchronized_data, Event.NO_MAJORITY
+
+        return None
```
Lots of code dup here.

Adamantios commented on scripts/test_prompts_tools.py:
What is this script for?

Adamantios replied:
What do you mean by "testing"? Would a notebook be more appropriate for this code?

Adamantios reviewed (approved):
LGTM, code will need to be refactored at some point to resolve some code dup, break huge files into multiple ones, big methods into smaller ones, etc.
